File: sirius-core/src/main/java/org/starcoin/core/ProtobufCodec.java
package org.starcoin.core;

import com.google.protobuf.GeneratedMessageV3;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.util.JsonFormat;

public interface ProtobufCodec<P extends GeneratedMessageV3> extends Protobufable<P> {

    P marshalProto();

    void unmarshalProto(P proto);

    @Override
    default P toProto() {
        return this.marshalProto();
    }


    default String toJson() {
        try {
            return JsonFormat.printer().print(this.toProto());
        } catch (InvalidProtocolBufferException e) {
            throw new RuntimeException(e);
        }
    }

}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/ProtobufCodec.kt
package org.starcoin.sirius.core

import com.google.protobuf.GeneratedMessageV3
import com.google.protobuf.InvalidProtocolBufferException
import com.google.protobuf.util.JsonFormat

interface ProtobufCodec<P : GeneratedMessageV3> : Protobufable<P> {

    fun marshalProto(): P

    fun unmarshalProto(proto: P)

    override fun toProto(): P {
        return this.marshalProto()
    }


    fun toJson(): String {
        try {
            return JsonFormat.printer().print(this.toProto())
        } catch (e: InvalidProtocolBufferException) {
            throw RuntimeException(e)
        }

    }

}
